From 4416afc79847e41e228c14add4efb1ed9db157dd Mon Sep 17 00:00:00 2001
From: Kihwal Lee <kihwal@apache.org>
Date: Mon, 2 Nov 2015 09:07:16 -0600
Subject: [PATCH 1195/2748] HDFS-9221.
 HdfsServerConstants#ReplicaState#getState should
 avoid calling values() since it creates a
 temporary array. (Staffan Friberg via yliu)
 (cherry picked from commit
 0ff1216100d16cfa862854a89cd1be8969b0bd7e)

(cherry picked from commit 2c51bbda510577506b3711c28b422b633ac59e5e)

Change-Id: If17c1501badac2ab6649f45b9cdabb192601d757
---
 .../hdfs/server/common/HdfsServerConstants.java    |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/common/HdfsServerConstants.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/common/HdfsServerConstants.java
index 3e8c842..9c2c8a9 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/common/HdfsServerConstants.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/common/HdfsServerConstants.java
@@ -249,6 +249,8 @@ public String toString() {
     /** Temporary replica: created for replication and relocation only. */
     TEMPORARY(4);
 
+    private static final ReplicaState[] cachedValues = ReplicaState.values();
+
     private final int value;
 
     private ReplicaState(int v) {
@@ -260,12 +262,12 @@ public int getValue() {
     }
 
     public static ReplicaState getState(int v) {
-      return ReplicaState.values()[v];
+      return cachedValues[v];
     }
 
     /** Read from in */
     public static ReplicaState read(DataInput in) throws IOException {
-      return values()[in.readByte()];
+      return cachedValues[in.readByte()];
     }
 
     /** Write to out */
-- 
1.7.9.5

